elfpy/bots/checkpoint_bot.py

```diff
@@ -0,0 +1,125 @@
+
+
+def get_config() -> EnvironmentConfig:
+    # Load some configuration variables from the environment.
+    load_dotenv()
+    artifacts_url = os.environ.get("ARTIFACTS_URL")
+    if artifacts_url is None:
+        raise ValueError("ARTIFACTS_URL environment variable must be set")
+    rpc_url = os.environ.get("RPC_URL")
+    if rpc_url is None:
+        raise ValueError("RPC_URL environment variable must be set")
+
+    # The configuration for the checkpoint bot halts on errors and logs to stdout.
+    return EnvironmentConfig(
+        # Networking
+        artifacts_url=artifacts_url,
+        rpc_url=rpc_url,
+        # Errors
+        halt_on_errors=True,
+        # Logging
+        log_stdout=True,
+        log_level=logging.INFO,
+    )
+
+
+def does_checkpoint_exist(hyperdrive: Contract, checkpointTime: int) -> bool:
+    return smart_contract_read(hyperdrive, "getCheckpoint", checkpointTime)["sharePrice"] > 0
+
+
+def main() -> None:
+    # Get the configuration and initialize the web3 provider.
+    config = get_config()
+    provider = eth.web3_setup.initialize_web3_with_http_provider(config.rpc_url, reset_provider=False)
+
+    # Setup logging
+    logs.setup_logging(
+        log_filename=config.log_filename,
+        max_bytes=config.max_bytes,
+        log_level=config.log_level,
+        delete_previous_logs=config.delete_previous_logs,
+        log_stdout=config.log_stdout,
+        log_format_string=config.log_formatter,
+    )
+
+    # Fund the checkpoint sender with some ETH.
+    balance = int(100e18)
+    sender = EthAccount(agent=Agent(wallet_address=0))
+    set_anvil_account_balance(provider, sender.account.address, balance)
+    logging.info(f"Successfully funded the sender={sender.account.address}.")
+
+    # Get the Hyperdrive contract.
+    hyperdrive_abis = eth.abi.load_all_abis(config.build_folder)
+    addresses = hyperdrive_interface.fetch_hyperdrive_address_from_url(
+        os.path.join(config.artifacts_url, "addresses.json")
+    )
+    hyperdrive: Contract = provider.eth.contract(
+        abi=hyperdrive_abis[config.hyperdrive_abi],
+        address=provider.to_checksum_address(addresses.mock_hyperdrive),
+    )
+
+    # Run the checkpoint bot. This bot will attempt to mint a new checkpoint
+    # every checkpoint after a waiting period. It will poll very infrequently
+    # to reduce the probability of needing to mint a checkpoint.
+    config = get_hyperdrive_config(hyperdrive)
+    checkpoint_duration = config["checkpointDuration"]
+    while True:
+        # Get the latest block time and check to see if a new checkpoint should
+        # be minted. This bot waits for a portion of the checkpoint to reduce
+        # the probability of needing a checkpoint. After the waiting period,
+        # the bot will attempt to mint a checkpoint.
+        timestamp = provider.eth.get_block("latest").timestamp  # type: ignore
+        checkpoint_portion_elapsed = timestamp % checkpoint_duration
```

Comment:
The `timestamp` variable is of type `Timestamp` which is a `NewType("Timestamp", int)`. Checkpoint duration is also an integer. [Integer modulo can behave differently from float](https://realpython.com/python-modulo-operator/#python-modulo-operator-basics). This has burned us in the past, so I just wanted to bring it to your attention. To be safe & consistent with elsewhere in the codebase, you might want to convert everything to `FixedPoint` before proceeding.

We also do intend to update the `hyperdrive_interface` functions to return FixedPoint for all values, instead of a mixture of floats and ints as it does now. The current behavior is from an oversight.